models/stock_picking_batch.py

```diff
@@ -309,37 +303,55 @@ def unpickable_item(self, reason, product_id=None, location_id=None, package_nam
 
         if picking_type_id is None:
             picking_type_id = ResUsers.get_user_warehouse().int_type_id.id
+
         if picking.batch_id != self:
             raise ValidationError(_('Move line is not part of the batch.'))
+
         if picking.state in ['cancel', 'done']:
             raise ValidationError(_('Cannot mark a move line as unpickable '
                                     'when it is part of a completed Picking.'))
 
+        original_picking_id = None
 
         if len(picking.move_line_ids - move_lines):
-            # Create a backorder for the affected move lines if there are
-            # move lines that are not affected
+            # Create a backorder for the affected move lines if
+            # there are move lines that are not affected
+            original_picking_id = picking.id
             picking = picking._create_backorder(move_lines)
 
-        # Remove the pick from the batch, and refine it
-        picking.batch_id = False
-        # By default the pick is cancelled
+        # By default the pick is unreserved
         picking._refine_picking(reason)
 
         group = Group.get_group(group_identifier=reason,
                                 create=True)
 
         # create new "investigation pick"
-        Picking.with_context(allow_partial=allow_partial).create_picking(quant_ids=quants.mapped('id'),
+        Picking.with_context(allow_partial=allow_partial)\
+               .create_picking(quant_ids=quants.mapped('id'),
                                location_id=location.id,
                                picking_type_id=picking_type_id,
                                group_id=group.id)
 
-        # If the batch does not contain any remaining picking to do, it can
-        # be set as done
+        # Try to re-assing the picking after, by creating the
+        # investigation, we've reserved the problematic stock
+        picking.action_assign()
+
+        # NB: if the product is not available, the picking will
+        # remain in the batch in 'waiting' state
+
+        if original_picking_id is not None and picking.state == 'assigned':
+            # A backorder has been created, but the stock is
+            # available; the move lines should be linked to the
+            # original picking to directly process the stock
+            picking.move_line_ids.write({'picking_id': original_picking_id})
+            picking.move_lines.write({'picking_id': original_picking_id})
+            picking.unlink()
+
+        # If the batch does not contain any remaining picking to do,
+        # it can be set as done
         remaining_pickings = self.picking_ids.filtered(
-            lambda x: x.state in ['assigned']
-        )
+            lambda x: x.state in ['assigned'])
```

Comment:
What happens if we have the backorder in state 'waiting'? Probably we shouldn't be finishing the batch.
Add 'waiting' ?

Reply:
Let me double check with the product fellows and get back to you.